### PR TITLE
Adding tests

### PR DESCRIPTION
**Pull request by JoeZiminski**
This PR adds first tests for configs, making directories and transferring data (on local filesystem). This also includes a PR (merged from adding_rclone into adding_tests) for depreciating pyftpsync and replacing with rclone.

Currently, SSH and remote transfer over a connection is not tested.

The main repo (not tests) is now typed and working with mypy setup on the pre-commit. GitHub actions also seems to be working well now

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -1,3 +1,49 @@
...
+name = "datashuttle"
+version = "0.0.0"
+description = "Organise and transfer scientific projects in BIDS format"
+readme = "README.md"
+requires-python = ">=3.9.0"
+
+authors = [
+    {name = "Joe Ziminski", email = "j.ziminski@ucl.ac.uk"},
+    {name = "Adam Tyson", email = "code@adamltyson.com"},
+    {name = "Niko Sirmpilatze", email = "n.sirmpilatze@ucl.ac.uk"},
```
Comment (niksirbi) [inline pyproject.toml:11]: better to use my github gmail for github/pypi things: niko.sirbiladze@gmail.com

File: datashuttle/configs.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+                "remote_path must contain the full directory path "
+                "with no ~ syntax"
+            )
+
+        # Check SSH settings
+        if self["ssh_to_remote"] is True and (
+            not self["remote_host_id"] or not self["remote_host_username"]
+        ):
+            utils.raise_error(
+                "ssh to remote set but no remote_host_id or "
```
Comment (niksirbi) [inline datashuttle/configs.py:53]: seems like a typo. Should it be 'but no remote_host_id is set or no remote_host_username is provided'?

**Comment by adamltyson** [comment]
FYI I've removed myself as a reviewer for this PR, but @JoeZiminski @niksirbi let me know if you'd like me to take a look at anything.

File: tests/test_configs.py [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+        assert project.cfg["ssh_to_remote"] is True
+
+    def test_no_ssh_options_set_on_make_config_file(self, project):
+        """
+        Check that program will assert if not all ssh options
+        are set on make_config_file
+        """
+        with pytest.raises(
+            BaseException
+        ) as e:  # TODO: checkk what the original exceptions were
```
Comment (niksirbi) [inline tests/test_configs.py:99]: typo in chekk

File: tests/test_utils.py [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+    on the  structure specified on project itself.
+
+    Cycle through all experiment type (defined in
+    project._ses_dirs()), sub, sessions and check that
+    the expected file exists. For  subdirs, recursively
+    check all exist.
+
+    Directories in which directory_used[key] (where key
+    is the cannoincal dict key in project._ses_dirs())
+    is not used are expected  not to be made, and this
```
Comment (niksirbi) [inline tests/test_utils.py:156]: there is a typo here. Also I don't really understand the sentence in this paragraph

File: tests/test_make_dirs.py [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+    def test_process_names_duplicate_ele(self, prefix):
+        """
+        Test that appropriate error is raised when duplicate name
+        is passed to process_names().
+        """
+        with pytest.raises(BaseException) as e:
+            utils.process_names(["1", "2", "3", "3", "4"], prefix)
+
+        assert (
+            "Subject and session names but all be unqiue "
```
Comment (niksirbi) [inline tests/test_make_dirs.py:68]: typo: must all be unique

File: tests/test_make_dirs.py [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+            "(i.e. there are no duplicates in list input)" == str(e.value)
+        )
+
+    def test_process_names_prefix(self, project):
```
Comment (niksirbi) [inline tests/test_make_dirs.py:72]: project is not used by this function

**Review by niksirbi — COMMENTED** [review]
Outstanding work and very meticulous on testing. I came across many things I din not know about, such as fixtures.

I left some sprinkled low-level comments here and there. Most of the high level stuff is already discussed and covered in other issues and branches.

A note on the `datashutttle.py` file, which currently has 1000 lines. Is there any way to refactor even more stuff out of there?

**Comment by niksirbi** [comment]
Also, when running `mypy -p datashuttle` on my machine, the following errors appear:
```python
datashuttle/utils_mod/utils.py:124: error: Incompatible default for argument "password" (default has type "None", argument has type "str")  [assignment]
datashuttle/utils_mod/utils.py:124: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
datashuttle/utils_mod/utils.py:124: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
datashuttle/utils_mod/utils.py:125: error: Incompatible default for argument "private_key_path" (default has type "None", argument has type "str")  [assignment]
datashuttle/utils_mod/utils.py:125: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
datashuttle/utils_mod/utils.py:125: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
datashuttle/datashuttle.py:127: error: Incompatible default for argument "ses_names" (default has type "None", argument has type "Union[str, List[Any]]")  [assignment]
datashuttle/datashuttle.py:127: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
datashuttle/datashuttle.py:127: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
datashuttle/datashuttle.py:302: error: Incompatible default for argument "remote_path_local" (default has type "None", argument has type "str")  [assignment]
datashuttle/datashuttle.py:302: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
datashuttle/datashuttle.py:302: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
datashuttle/datashuttle.py:303: error: Incompatible default for argument "remote_path_ssh" (default has type "None", argument has type "str")  [assignment]
datashuttle/datashuttle.py:303: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
datashuttle/datashuttle.py:303: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
datashuttle/datashuttle.py:304: error: Incompatible default for argument "remote_host_id" (default has type "None", argument has type "str")  [assignment]
datashuttle/datashuttle.py:304: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
datashuttle/datashuttle.py:304: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
datashuttle/datashuttle.py:305: error: Incompatible default for argument "remote_host_username" (default has type "None", argument has type "str")  [assignment]
datashuttle/datashuttle.py:305: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
datashuttle/datashuttle.py:305: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
```
I think these have to do with the fact that you do not include `None` as a possible type in the `Union`, and then you declare `None` to be the default value.

**Review by niksirbi — APPROVED** [review]

**Review by niksirbi — APPROVED** [review]

**Comment by JoeZiminski** [comment]
Thanks a lot Niko, very good point about datashuttle.py I will make an issue to do this. I'll merge for now and adress these on a newer PR as I think some of that code has gone.

**Pull request merged**